File: plugins/maven/src/test/java/org/jetbrains/idea/maven/project/importing/MavenProjectsTreeTestCase.java
/*
 * Copyright 2000-2016 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.jetbrains.idea.maven.project.importing;

import com.intellij.maven.testFramework.MavenMultiVersionImportingTestCase;
import com.intellij.openapi.application.WriteAction;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.Pair;
import com.intellij.openapi.vfs.VirtualFile;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.idea.maven.model.MavenExplicitProfiles;
import org.jetbrains.idea.maven.project.*;
import org.jetbrains.idea.maven.server.NativeMavenProjectHolder;
import org.jetbrains.idea.maven.utils.MavenProcessCanceledException;
import org.jetbrains.idea.maven.utils.MavenProgressIndicator;
import org.jetbrains.idea.maven.utils.MavenUtil;

import java.io.IOException;
import java.util.Collections;
import java.util.List;
import java.util.Set;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.stream.Collectors;

import static java.util.Arrays.asList;

public abstract class MavenProjectsTreeTestCase extends MavenMultiVersionImportingTestCase {
  protected MavenProjectsTree myTree;


  @Override
  protected void setUpInWriteAction() throws Exception {
    super.setUpInWriteAction();
    myTree = MavenProjectsManager.getInstance(myProject).getProjectsTree();
  }

  protected void updateAll(VirtualFile... files) {
    updateAll(Collections.emptyList(), files);
  }

  protected void updateAll(List<String> profiles, VirtualFile... files) {
    myTree.resetManagedFilesAndProfiles(asList(files), new MavenExplicitProfiles(profiles));
    myTree.updateAll(false, getMavenGeneralSettings(), getMavenProgressIndicator().getIndicator());
  }

  protected void update(VirtualFile file) {
    myTree.update(asList(file), false, getMavenGeneralSettings(), getMavenProgressIndicator().getIndicator());
  }

  protected void deleteProject(VirtualFile file) {
    myTree.delete(asList(file), getMavenGeneralSettings(), getMavenProgressIndicator().getIndicator());
  }

  protected void updateTimestamps(final VirtualFile... files) throws IOException {
    WriteAction.runAndWait(() -> {
      for (VirtualFile each : files) {
        each.setBinaryContent(each.contentsToByteArray());
      }
    });
  }

  protected static ListenerLog log() {
    return new ListenerLog();
  }

  protected static class ListenerLog extends CopyOnWriteArrayList<Pair<String, Set<String>>> {
    ListenerLog() { super(); }

    ListenerLog(ListenerLog log) { super(log); }

    ListenerLog add(String key, String... values) {
      var log = new ListenerLog(this);
      log.add(new Pair<>(key, Set.of(values)));
      return log;
    }
  }

  protected static class MyLoggingListener implements MavenProjectsTree.Listener {
    List<Pair<String, Set<String>>> log = new CopyOnWriteArrayList<>();

    private void add(String key, Set<String> value) {
      log.add(new Pair<>(key, value));
    }

    @Override
    public void projectsUpdated(@NotNull List<Pair<MavenProject, MavenProjectChanges>> updated, @NotNull List<MavenProject> deleted) {
      append(MavenUtil.collectFirsts(updated), "updated");
      append(deleted, "deleted");
    }

    private void append(List<MavenProject> updated, String text) {
      add(text, updated.stream().map(each -> each.getMavenId().getArtifactId()).collect(Collectors.toSet()));
    }

    @Override
    public void projectResolved(@NotNull Pair<MavenProject, MavenProjectChanges> projectWithChanges,
                                NativeMavenProjectHolder nativeMavenProject) {
      add("resolved", Set.of(projectWithChanges.first.getMavenId().getArtifactId()));
    }

    @Override
    public void pluginsResolved(@NotNull MavenProject project) {
      add("plugins", Set.of(project.getMavenId().getArtifactId()));
    }

    @Override
    public void foldersResolved(@NotNull Pair<MavenProject, MavenProjectChanges> projectWithChanges) {
      add("folders", Set.of(projectWithChanges.first.getMavenId().getArtifactId()));
    }
  }

  protected void resolve(@NotNull Project project,
                         @NotNull MavenProject mavenProject,
                         @NotNull MavenGeneralSettings generalSettings,
                         @NotNull MavenEmbeddersManager embeddersManager,
                         @NotNull MavenConsole console,
                         @NotNull MavenProgressIndicator process) throws MavenProcessCanceledException {
    var resolver = MavenProjectResolver.getInstance(project);
    resolver.resolve(List.of(mavenProject),
                     myTree,
                     generalSettings,
                     embeddersManager,
                     console,
                     process.getIndicator(),
                     process.getSyncConsole());
  }
}

File: plugins/maven/src/test/java/org/jetbrains/idea/maven/project/importing/MavenProjectsTreeTestCase.kt
// Copyright 2000-2023 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.idea.maven.project.importing

import com.intellij.maven.testFramework.MavenMultiVersionImportingTestCase
import com.intellij.openapi.application.WriteAction
import com.intellij.openapi.components.service
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.Pair
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.platform.util.progress.RawProgressReporter
import com.intellij.testFramework.RunAll
import com.intellij.util.ThrowableRunnable
import org.jetbrains.idea.maven.buildtool.MavenLogEventHandler
import org.jetbrains.idea.maven.model.MavenExplicitProfiles
import org.jetbrains.idea.maven.project.*
import org.jetbrains.idea.maven.utils.MavenUtil
import java.io.IOException
import java.util.concurrent.CopyOnWriteArrayList

abstract class MavenProjectsTreeTestCase : MavenMultiVersionImportingTestCase() {
  private var myTree: MavenProjectsTree? = null
  protected val rawProgressReporter: RawProgressReporter = object : RawProgressReporter {}
  protected lateinit var mavenEmbedderWrappers: MavenEmbedderWrappers

  val tree: MavenProjectsTree
    get() {
      return myTree!!
    }

  @Throws(Exception::class)
  override fun setUpInWriteAction() {
    super.setUpInWriteAction()
    myTree = MavenProjectsManager.getInstance(project).getProjectsTree()
  }

  override fun setUp() {
    super.setUp()
    mavenEmbedderWrappers = project.service<MavenEmbedderWrappersManager>().createMavenEmbedderWrappers()
  }

  override fun tearDown() {
    RunAll(
      ThrowableRunnable { mavenEmbedderWrappers.close() },
      ThrowableRunnable { super.tearDown() }
    ).run()
  }

  protected suspend fun updateAll(vararg files: VirtualFile) {
    updateAll(emptyList<String>(), *files)
  }

  protected suspend fun updateAll(profiles: List<String?>?, vararg files: VirtualFile) {
    tree.resetManagedFilesAndProfiles(listOf(*files), MavenExplicitProfiles(profiles))
    tree.updateAll(false, mavenGeneralSettings, mavenEmbedderWrappers, rawProgressReporter)
  }

  protected suspend fun update(file: VirtualFile) {
    tree.update(listOf(file), false, mavenGeneralSettings, mavenEmbedderWrappers, rawProgressReporter)
  }

  protected suspend fun deleteProject(file: VirtualFile) {
    tree.delete(listOf(file), mavenGeneralSettings, mavenEmbedderWrappers, rawProgressReporter)
  }

  @Throws(IOException::class)
  protected fun updateTimestamps(vararg files: VirtualFile) {
    WriteAction.runAndWait<IOException> {
      for (each in files) {
        each.setBinaryContent(each.contentsToByteArray())
      }
    }
  }

  class ListenerLog : CopyOnWriteArrayList<Pair<String, Set<String>>> {
    internal constructor() : super()
    internal constructor(log: ListenerLog) : super(log)

    fun add(key: String, vararg values: String): ListenerLog {
      val log = ListenerLog(this)
      log.add(Pair<String, Set<String>>(key, setOf(*values)))
      return log
    }
  }

  protected class MyLoggingListener : MavenProjectsTree.Listener {
    @JvmField
    var log: MutableList<Pair<String, Set<String?>>> = CopyOnWriteArrayList()
    private fun add(key: String, value: Set<String?>) {
      log.add(Pair(key, value))
    }

    override fun projectsUpdated(updated: List<Pair<MavenProject, MavenProjectChanges>>, deleted: List<MavenProject>) {
      append(MavenUtil.collectFirsts(updated), "updated")
      append(deleted, "deleted")
    }

    private fun append(updated: List<MavenProject>, text: String) {
      add(text, updated.map { it.mavenId.artifactId }.toSet())
    }

    override fun projectResolved(projectWithChanges: Pair<MavenProject, MavenProjectChanges>) {
      add("resolved", setOf(projectWithChanges.first.mavenId.artifactId))
    }

    override fun pluginsResolved(project: MavenProject) {
      add("plugins", setOf(project.mavenId.artifactId))
    }

    override fun foldersResolved(projectWithChanges: Pair<MavenProject, MavenProjectChanges>) {
      add("folders", setOf(projectWithChanges.first.mavenId.artifactId))
    }
  }

  protected suspend fun resolve(project: Project,
                                mavenProject: MavenProject,
                                generalSettings: MavenGeneralSettings
  ) {
    val resolver = MavenProjectResolver(project)
    val progressReporter = object : RawProgressReporter {}
    val updateSnapshots = projectsManager.forceUpdateSnapshots || generalSettings.isAlwaysUpdateSnapshots
    resolver.resolve(true,
                     listOf(mavenProject),
                     tree,
                     tree.workspaceMap,
                     generalSettings.effectiveRepositoryPath,
                     updateSnapshots,
                     mavenEmbedderWrappers,
                     progressReporter,
                     MavenLogEventHandler)
  }

  companion object {
    @JvmStatic
    protected fun log(): ListenerLog {
      return ListenerLog()
    }
  }
}
